[Persistence] Persist on mutation

VWoeltjen opened the pull request:
Refactoring for #825; remove need to explicitly persist after mutation by listening for mutation events and persisting when these occur. Opening early for early feedback.

Still to-do:

- [x] Update any test cases
- [x] Self-test touched behaviors
  - [x] "Link" action (actually drag-drop compose)
  - [x] Edit Properties
  - [x] Remove
  - [x] EditRepresenter (esp. `$scope.commit()`)
  - [x] Move action (uses LocationCapability.setPrimaryLocation)
  - [x] Link action
  - [x] Timer start/restart actions
  - [x] Timeline editing (horizontal dragging)
  - [x] Timeline editing (setting activity modes, links)
  - [x] Timeline editing (swimlane drag-drop)
  - [x] Editing of newly-created objects
- [x] Satisfy JSHint

VWoeltjen commented:
### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

akhenry commented on platform/core/src/runs/PersistingMutationListener.js:
```diff
@@ -0,0 +1,42 @@
+
+define([], function () {
+    /**
+     * Listens for mutation on domain objects and triggers mutation when
```
Listens for mutation on domain objects and triggers ~~mutation~~ persistence

VWoeltjen replied:
Fixed in fb8fe935292af7eca43bd2a95605f56090239b4c

akhenry commented:
I wonder if we should stop and re-assess our end goal here. The driving mantra is "mutate mutates and persist persists", but I'm not sure if this gets us any closer? 

This introduces a listener on mutation that automatically persists, so now mutate...persists, and persist persists, except if it's in a transaction. 

| Action        | Result (Before #874)                      | Result (after #874)                   |
| ----------------- |:------------------------------------------:|:----------------------------------------:|
| __Mutate__  | mutates                                       | mutates + persists                    | 
| __Persist__  | persists (except in transactions) | persist (except in transactions) |

I suppose on the plus side, this approach removes the decision of when to persist from the actions, making them more flexible, but it also introduces ambiguity for developers.

I think that ideally mutation should mutate with no unexpected side effects (such as persistence), and that persist should be called when you mean to persist an object. Ideally, in an editing scenario, there would be only one persist() call, made by the SaveAction. In our architecture however, actions should sometimes persist immediately (outside of edit mode) and sometimes not (when in edit mode). My thinking is that persistence should be removed from actions, and deferred to the the calling context (eg. context menus). This PR gets us some of the way there. How much extra work would it be to just go all the way, and call persist after actions are invoked and persistence is required?

VWoeltjen commented:
> I think that ideally mutation should mutate with no unexpected side effects (such as persistence), and that persist should be called when you mean to persist an object. Ideally, in an editing scenario, there would be only one persist() call, made by the SaveAction. In our architecture however, actions should sometimes persist immediately (outside of edit mode) and sometimes not (when in edit mode). My thinking is that persistence should be removed from actions, and deferred to the the calling context (eg. context menus). This PR gets us some of the way there. How much extra work would it be to just go all the way, and call persist after actions are invoked and persistence is required?

I see your point. I see a few immediate problems with handling this from the calling context of actions:

* Actions aren't the only place where mutation occurs or might occur. This could lead to confusion (I'm writing actions and I don't need to think about persistence, but introduce some in-view editing and now I do. Why?)
* Actions may need to persist more than just the object they were initiated from (Remove modifies the parent, for instance.) Tracking which objects were made dirty by which actions is non-trivial because of this, especially if actions involve asynchrony.
* We still carry the "persist persists, except in edit mode" ambiguity, and don't have a good way to move away from it (since non-action feature code would still need to call persist, but we would likely not want this to actually persist in Edit mode.)

One approach I see to resolving this would be to actually introduce a Mode abstraction (it's conspicuous that, as much as we talk about Browse mode versus Edit mode, there is not EditMode or BrowseMode in the code.) While BrowseMode is active, it listens for mutation and persists immediately. While EditMode is active, it listens for mutation and puts things in transactions (replacing the transactional persistence decorator). In this situation, "mutate mutates and persist persists" and any additional side effects are strictly behaviors of the Mode.

akhenry commented:
> Actions aren't the only place where mutation occurs or might occur. This could lead to confusion (I'm writing actions and I don't need to think about persistence, but introduce some in-view editing and now I do. Why?)

I totally agree, the aim is to remove ambiguity here. Which is why I think the decision to persist should be in the hands of the developer, not the framework.

We need to be clear about what kind of modifications occur in edit mode, and which ones are reflected immediately, otherwise "edit mode" is a very confusing paradigm. (Maybe we don't need edit mode at all, persisting any modifications immediately reduces the likelihood of collisions with changes by other users. That's a much bigger discussion though :) )

Does in-view editing happen outside of an edit context? If so, should it? If it's a persistent change, then maybe it should only happen in edit mode. If it's not persisted, then no problem, no need to call persist in either case. 

Agreed, there is no application-wide 'edit mode' or 'browse mode' state, and I too have wondered if we should have one. In the cases where there were ambiguity about whether or not to persist, it might be necessary. We probably have enough information already from `inEditContext` to make this decision though. eg. the ContextMenuGesture could check edit mode, and persist as appropriate at that point, after the action is completed. We would, of course, be baking edit mode into the framework by doing that, but it is a fairly essential piece of functionality right now.

larkin commented:
Okay-- I'm just reading discussion here, haven't reviewed code.  As I understand it, I don't see any ambiguity here about persistence.

Actions don't persist-- they mutate.  The application decides whether or not to persist changes immediately (browse mode) or sandbox all mutations and allow for discarding or committing (as in edit mode).  The point here is the actions (and their developers) don't need to think about persistence... so they don't.

I do see that if developers aren't taking persistence into consideration, then they have no way of handling persistence failures in browse mode.  Is this the problem you see?  In this case, we wrap the execution of a context menu action in a transaction, and handle success/fail.  We already have bugs with asynchronous action execution which this would also cause us to fix, e.g. #963.

At baseline, we have poorly defined the unit of work accomplished by actions (actions can include other actions) and unfortunately confused it with UX concerns (some actions are displayed to users), and as a result, we have this ambiguity.  Separating actions from UX concerns should solve for the ambiguity.

larkin commented:
I posted that comment prematurely and just significantly rewrote it.  If you were previously responding to it; reread it before going further!

VWoeltjen commented:
> Okay-- I'm just reading discussion here, haven't reviewed code. As I understand it, I don't see any ambiguity here about persistence. I see ambiguity about what an action is.

> In one case, an action a unit of work; a building block for more complex actions. In another case, it's an asynchronous action that may succeed or fail. In other cases, it's synchronous. In all of these cases the action may or may not be user facing.

I agree that this ambiguity exists; I've likened this to the "plumbing" versus "porcelain" of git commands. I think the pattern used by the entanglement actions (where the "plumbing" is handled by services) is a good one.

> I think we need to fix this:

> as a small, contained, reusable unit of work, actions don't persist-- they mutate. They are verbs for usage by developers.

> Context actions and other actions available in browse mode are special, user-facing, asynchronous, complex sets of actions that have pending, success, and failure states. If these combine multiple actions they should initiate a transaction, perform their changes, and persist them, so that they can handle success/failure gracefully.

I'll frame my comments in terms of "Actions" (seen by users) and "Verbs" (used by developers):

* Actions have a `perform()` method and sufficient metadata to display to the user.
  * Actions, as a matter of convention, generally just sequence some series of verbs.
  * Actions are responsible for triggering persistence, if that's part of the action
  * Actions _may_ trigger other actions, but that should not be the preferred style
    * _I think this could make policy management difficult, but the existing paradigm isn't all that effective in practice either_
* Verbs are just functions that are made available _somewhere_ (e.g. as services)
  * As a convention, verbs don't persist objects, they mutate them
  * There may be exceptions to this, but they'd be clearly documented. (e.g. "persist" might be one such verb)

Some of the challenges with this approach:

1. If actions invoke a "persist" method, that's still confusing; we want actions to be available in Edit mode, but we don't want them to persist immediately. (I think the above still violates the "persist persists" goal because of this.)
2. Actions do not necessarily know which objects were mutated by their verbs. Relying on prior knowledge ("after I use the verb _remove_, I need to use the verb _persist_ on the object's parent") makes it hard to change the implementation of those verbs (reducing their value)
3. If actions start a transaction (e.g. to chain multiple actions) during Edit mode, that's problematic - there's already a transaction active.

Possible solutions, numbered as above:

1. Don't call it "persist", call it something else ("commit"?) which is closer to the actual meaning: Mark these changes as ready to persist. (More specifically, persist these changes at the end of the active transaction, or immediately if there is no active transaction; but the developer shouldn't be thinking about that level of detail anyway. It's inconsistent with our UI style to have actions which literally persist, except for things like Save.)
2. As a convention, verbs could return a promise for an array of domain objects that have been modified. (This is tough with verbs that we want to return other objects - for instance, it's still useful for `composition.add` to return the domain object in its new context.)
  * Alternative: Verbs take responsibility for commit calls too.
  * Alternative: Actions pass a transaction (which could be as simple as an array) into verbs, which can add to that transaction ("these are the things to persist")
3. Transactions could "stack" (e.g. opening and finally completing a transaction from an action in Edit mode would be allowable, but it wouldn't actually persist until the top-level transaction initiated by Edit mode was ended.) This could behave really unexpectedly, though, if two transactions were started and ended concurrently (which, since actions are often promise-chainy, is very plausible.)
  * Alternative: Don't return promises from mutation, or from verbs in general. Most transactions should resolve synchronously; Edit mode is the one exception.
  * Alternative: Use a "stacked" form of transactions, but return an object that is used to end the transaction (instead of using a method on transactionService). You still have the problem of transactions appearing "nested" when they are concurrent, but at least you don't have to worry about closing the wrong one.
  * Alternative: Pass transaction objects into the `perform` method of actions; these may be passed in to other actions as well. In Browse mode, this transaction gets completed at the end of the action; in Edit mode, this transaction gets shared (or is used to make "subtransactions", since an action might conceivably cancel its own transaction) until Save is hit.

VWoeltjen commented:
Returning to this discussion; had a thought that links together some [other](https://github.com/nasa/openmct/pull/1084#issuecomment-233034627) [discussions](https://github.com/nasa/openmct/pull/1074) we've been having. What if mutation was part of a Transaction, and persistence was just an implicit consequence?

```
var transaction = transactionService.newTransaction();
transaction.mutate(domainObject, doSomeMutating);
transaction.mutate(parentObject, moreMutating);
transaction.create(newObject);
transaction.commit();
```

Does `transaction.commit` persist? Depends on the mode. As plugin developers, we just don't think about persistence. We just say "here's a set of related changes" and leave it to the platform to decide what to do with those.

Using "verbs" versus "actions" as defined in previous comments, there's a decision about where transactions take place. I think that, as a convention, verbs should allow a transaction to be passed in (perhaps as an optional argument, creating/committing a new transaction if none is supplied?) There are still a lot of forms this could take, but I think in general it makes sense if actions start/commit transactions, and verbs normally add to transactions.

This also means we can fire mutation callbacks a bit more smartly (e.g. when commit is called, instead of when mutate is called) and gives us better support for Undo (assuming each transaction knows how to unwind itself.) It even points toward, dare I say it, a merge strategy down the road (replay transaction upon the latest)

Also implied is that, somewhere, there is a `persist` verb that does exactly what it says. It gets invoked as a consequence of `commit`, depending on the mode (meaning that modes would need to be able to dictate how transactions are handled while they are in effect.) There's work to be done on getting those specifics right.

Benefits of this approach:

* No magic; no little demon hiding in the background watching for mutate events and triggering persistence.
* Transparent; the different persistence behaviors of Browse and Edit mode are well-hidden, without any hijacked semantics (by saying "commit" instead of "persist")
* "Mutate mutates and persist persists": Yep! With the caveat that mutation is now clearly in the context of a transaction; a change, but I think a helpful one.

Thoughts? It's a significant change but I think it offers significant simplification (more total concepts, true, but the concepts become simpler/clearer)

larkin commented:
> @VWoeltjen 
> What if mutation was part of a Transaction, and persistence was just an implicit consequence?

I like persistence as a consequence of a transaction, but I don't like mutation only being available within a transaction, as anything that wants to mutate suddenly has to deal with transactions, and has to deal with the asynchronous nature of them.

> @VWoeltjen 
> This also means we can fire mutation callbacks a bit more smartly (e.g. when commit is called, instead of when mutate is called) and gives us better support for Undo (assuming each transaction knows how to unwind itself.) It even points toward, dare I say it, a merge strategy down the road (replay transaction upon the latest)

This is backwards from our overall design goals, it directly prevents us from synchronizing different views via mutation events-- another reason why mutation shouldn't explicitly be part of a transaction.

The original idea was to use transactions as a way to prevent having to wrap domain objects with all this crazy functionality, and thus to remove editable domain objects and clean up code.  

We went down a long path with the "persist after mutate" magic and I think after discussion we have found problems with that design and we don't want to go through with it.

I think you touched on a great alternative in your original response, to restate:

1. We shouldn't call it persist, it's a transaction and it should be called "commit".  Commit is "all pass or all fail"; any single persistence failure in a transaction should undo any successful persistence calls inside that transaction.  An error handler could be used to retry the transaction based on user input or other algorithm, or just allow it to fail and roll back the entire application state.
2. Transactions can be nested.  (you said "stacked").  So edit mode starts a transaction, and actions themselves may have transactions within the edit mode transaction, and actions may utilize other actions which themselves have transactions.  Nothing is persisted until the root transaction is committed.  Nested transactions could attempt to recover from their own errors, and if they fail to do so the parent transaction would fail, and everything would then be rolled back.

As follows from those ideas: transactions should hook into the mutation service and track:

1. a list of dirty objects
2. the initial state of those dirty objects 
3. the current (mutated) state of those dirty objects

When a transaction is committed, it would persist the current state of all dirty objects and track those that persist successfully.  If any of them fail to persist, it would immediately re-persist the initial version of any successfully persisted objects and then throw an error.

This would solve actions in edit mode and nested actions.  

The only question I have is, what happens when you mutate something without an active transaction?  Personally, I think mutating something without an active transaction should toggle edit mode which would in turn start a new transaction.  Users can experimentally change things but you always want that rollback option.

VWoeltjen commented:
> I like persistence as a consequence of a transaction, but I don't like mutation only being available within a transaction, as anything that wants to mutate suddenly has to deal with transactions, and has to deal with the asynchronous nature of them.

Fair point, that does allow a lot of concerns to piggy-back on the simple task of changing some state.

Trying to summarize (for my own benefit, and as a check for understanding):

MutableObject, from #1074: Provides methods to mutate and observe mutation of domain objects. Makes no distinctions about modes; has no concept of transactions. Just changes state of domain objects and broadcasts these changes to allow (but not provide) synchronization.

TransactionService: Allows new transactions to be started. There is always one "active" transactions (or maybe zero in the Browse case?); starting a new transaction makes a transaction the child of the active transaction, then promotes it to be the active transaction.

Transaction: Has a `commit` and a `cancel` method; when you start a new transaction, you get one of these you can commit later. (This also seems like a reasonable place to attach listeners for what to "do" when a transaction is committed/cancelled; or those could be passed into the TransactionService when a transaction is started.)

(Some intermediary): Listens for mutation events to add dirty objects to the active transaction.

EditMode: Starts a new transaction when it's started, commits the transaction on Save, cancels the transaction on Cancel.

> The only question I have is, what happens when you mutate something without an active transaction? Personally, I think mutating something without an active transaction should toggle edit mode which would in turn start a new transaction. Users can experimentally change things but you always want that rollback option.

One thing I like about this approach, architecturally, is that it provides some flexibility for handling this case (no active transaction.) For our current behavior, we could create a new transaction, add to it, and persist it immediately. For the behavior you describe, we could swap out or reimplement that piece, without touching the rest.

VWoeltjen commented:
![nomnoml-9](https://cloud.githubusercontent.com/assets/1685764/17148602/b2c20fba-531c-11e6-9752-ce5985ecf824.png)

* TransactionService would be responsible for ensuring that stacked/nested transactions are wired such that they appropriately commit/cancel each other. Not sure if there should be a `rollback` separate from `cancel`
* Need to decide on conventions for promises. I think both `commit` and `cancel` should reject if any of their associated callbacks throw or reject, and resolve otherwise. Need to make sure that "if commit fails, invoke cancel behavior" cascades appropriately

VWoeltjen commented:
@larkin / @akhenry - switched to the approach described above. Difference from last review is fairly small:

* Refactored Transaction out of TransactionService
* Added support for nested transactions
* Changed the mutation listener to add to transactions instead of issuing persist calls

At this point, no actions are directly using transactions (although some could be), nor is there anything starting and ending transactions around `action.perform()` calls. I'm thinking these are out-of-scope for this issue unless it would be helpful to validate this approach.

larkin commented:
I like this new implementation, it simplifies things quite nicely.

VWoeltjen commented:
Groovy! Self-assigning to finish up rests, will add a new review checklist when that is done

VWoeltjen commented:
Finished adding tests and JSDoc for latest changes here. Ready to merge once we exit code freeze. The commit history is long, but still fairly consistently incremental in nature, so haven't squash-merged/rebased to simplify.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

akhenry commented:
I like this - reframing what we're doing as 'committing' rather than persisting makes sense. I have a couple of thoughts:

* I'm not entirely sure why we're removing the need to call persist. I know that was the goal of the original issue, but I don't recall why. From memory it was in order to remove `persist()` calls from actions, with the goal being to remove ambiguity about what `persist()` does. I think we achieve this by effectively replacing it in most cases with `commit()` though, and I wonder if we should just stop there. Automatically persisting any mutation outside of edit mode removes some flexibility and adds some mystery to the API ("How are my objects getting persisted?"). We should be sure we want to do this. 

Perhaps for now we might be better with an approach where we retained commit() calls in the actions, in place of the current persist() calls? 

* Exposing both 'commit' and persist may be confusing to plugin developers. If we do this, we should be sure to clarify their roles in docs and examples. That said, by automatically persisting mutations, and handling

akhenry commented on platform/commonUI/edit/src/services/NestedTransaction.js:
```diff
@@ -0,0 +1,48 @@
+     * @extends {platform/commonUI/edit/services.Transaction}
+     * @memberof platform/commonUI/edit/services
+     */
+    function NestedTransaction(parent) {
```
Wondering if the NestedTransaction class is necessary. If we add a parent to the `Transaction` class then `NestedTransaction` could perhaps be replaced with a simple if statement in the existing commit/cancel?

eg.

``` javascript
['commit', 'cancel'].forEach(function (method) {
    Transaction.prototype[method] = function () {
        var promises = [];
        var callback;
  
        while (this.callbacks.length > 0) {
            callback = this.callbacks.shift();
            //If nested, add to parent to be committed later
            if (this.parent){
                this.parent.add(callback.commit, callback.cancel);
            } else {
                try {
                    promises.push(callback[method]());
                } catch (e) {
                    this.$log
                        .error("Error trying to " + method + " transaction.");
                }
            }
        }
        return Promise.all(promises);
    };
});
```

VWoeltjen replied:
I see your point, but considering the logic in play I think iffing around the presence of `parent` makes the logic just a tad harder to follow and reason about the difference between the base case and the general case here. Separating this out into two classes does add some boilerplate, which is not great, but does seem a little clearer

akhenry replied:
@VWoeltjen Makes sense, happy for it to stay as two classes.

akhenry commented on platform/commonUI/edit/src/services/TransactionService.js:
```diff
@@ -37,10 +37,8 @@ define(
-
-            this.onCommits = [];
-            this.onCancels = [];
+            this.transaction = undefined;
```
I'd be inclined to do away with this variable and just use `this.transactionStack[this.transactionStack.length() - 1]`, it's one less thing to maintain and it can potentially be inconsistent with the state of the stack

larkin commented:
@akhenry We are effectively replacing `persist` with `commit` so that developers do not need to worry about whether the application is in edit mode (where all changes are sandboxed until the user saves them), while allowing them to ensure a transaction will either fully succeed or fully fail.  Developers don't particularly care about the inner workings of the persist calls, but they want some guarantee of ACID properties.

> Perhaps for now we might be better with an approach where we retained commit() calls in the actions, in place of the current persist() calls?

I agree here, in that all actions that mutate should be expected to handle asynchronous success/error callbacks as well as the "pending" state while persist operations are being completed.

larkin commented:
@akhenry to take another look.

VWoeltjen commented:
Addressed comments from code review, believe this is ready to merge now.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

akhenry commented:
As much as I want to close out this PR (and I really, really do), persistence as an automatic consequence of mutation makes me a little uneasy. 

@VWoeltjen do you have any thoughts on https://github.com/nasa/openmct/pull/874#issuecomment-243086862 ? Mostly this suggestion, which was seconded by @larkin -

> Perhaps for now we might be better with an approach where we retained commit() calls in the actions, in place of the current persist() calls?

Don't make any changes right away, would just like to hear thoughts.

larkin commented:
I'm okay with persisting as a result of mutation w/o any changes.  We have a lot of in-progress discussions on the actions API.

akhenry commented:
Ok, in that case I'm happy to go with majority rather than unanimous agreement.

### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
